sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -102,6 +104,9 @@ object AnsiTypeCoercion extends TypeCoercionBase {
     case (NullType, t1) => Some(t1)
     case (t1, NullType) => Some(t1)
 
+    case(s1: StringType, s2: StringType) if SQLConf.get.preserveCharVarcharTypeInfo =>
```

Comment:
If we see char/varchar in the type system, it means `preserveCharVarcharTypeInfo` is true and we don't rewrite them to StringType at the beginning. We don't need to check the conf again here.